src/metabase/util/log/capture.cljc

```diff
@@ -0,0 +1,205 @@
+     that starts with `metabase.`
+
+  3. A vector of [namespace level], e.g. `[my.namespace :debug]`, which would capture all `:debug` and `:trace`
+     messages inside `my.namespace`, `my.namespace.x`, etc.
+
+  A code example is worth a thousand-word docstring, so here is an example
+
+  (log.capture/with-log-messages-for-level [messages [metabase.util.log.capture-test :trace]]
+    (is (= []
+           (messages)))
+    (log/trace \"message\")
+    (is (= [{:namespace 'metabase.util.log.capture-test
+             :level     :trace
+             :e         nil
+             :message   \"message\"}]
+           (messages))))
+
+  See [[metabase.util.log.capture-test]] for more example usages.
+
+  You can pass multiple bindings to this macro without them affecting one another, regardless of whether the things
+  they capture overlap or not. See [[metabase.util.log.capture-test/multiple-captures-test]] for an example."
+  {:arglists '([[messages-fn-binding ns-level & more-bindings] & body])}
+  [& args]
+  (let [{:keys [bindings body]} (s/conform ::with-log-messages-for-level-args args)]
+    (reduce
+     (fn [form bindings]
+       (let [{:keys [messages-fn-binding ns-level]} bindings
+             [ns-level-type ns-level]               ns-level
+             {:keys [ns-str level]}                 (case ns-level-type
+                                                      :ns-level ns-level
+                                                      :ns       {:ns-str ns-level, :level :trace}
+                                                      :level    {:ns-str "metabase", :level ns-level})]
+         `(do-with-log-messages-for-level ~(str ns-str) ~(level->int level) (fn [~messages-fn-binding] ~form))))
+     `(do ~@body)
+     bindings)))
+
+(s/fdef with-log-messages-for-level
+  :args ::with-log-messages-for-level-args
+  :ret  any?)
+
+;;; The macroexpansion of something like
+;;;
+;;;    (log/trace "a picture")
+;;;
+;;; becomes
+;;;
+;;;    (do
+;;;      (metabase.util.log.capture/capture-logp
+;;;       "metabase.util.log.capture-test"
+;;;       :trace
+;;;       "a picture")
+;;;      ;; clojure.tools.logging stuff
+;;;      ...)
+;;;
+;;; becomes
+;;;
+;;;    (do
+;;;      (when-let [f# (metabase.util.log.capture/*capture-logs-fn*
+;;;                     "metabase.util.log.capture-test"
+;;;                     5)]
+;;;          (metabase.util.log.capture/capture-logp! f# "a picture"))
+;;;      ...)
+;;;
+;;; VERY VERY IMPORTANT! [[*capture-logs-fn*]] only returns a function if logs for that namespace AND level are to be
+;;; captured! Only then is [[capture-logp!]] or [[capture-logf!]], which ultimately call the returned function `f#`,
+;;; called with the args to the logging macro -- and only then do they get evaluated! It's super important not to
+;;; evaluate the log args unless we actually want to capture them! They might be very expensive!
```

Comment:
Hmmm, why expensive? Is this about `print-str` and overall string generation or more?